src/serac/physics/thermal_conduction_functional.hpp

```diff
@@ -384,6 +487,79 @@ class ThermalConductionFunctional : public BasePhysics {
     }
   }
 
+  /**
+   * @brief Solve the adjoint problem
+   * @pre It is expected that the forward analysis is complete and the current temperature state is valid
+   * @note If the essential boundary state is not specified, homogeneous essential boundary conditions are applied
+   *
+   * @param[in] adjoint_load The dual state that contains the right hand side of the adjoint system
+   * @param[in] dual_with_essential_boundary A optional finite element dual containing the non-homogenous essential
+   * boundary condition data for the adjoint problem
+   * @return The computed adjoint finite element state
+   */
+  virtual const serac::FiniteElementState& solveAdjoint(FiniteElementDual& adjoint_load,
+                                                        FiniteElementDual* dual_with_essential_boundary = nullptr)
+  {
+    adjoint_load.initializeTrueVec();
+
+    // note: The assignment operator must be called after the copy constructor because
+    // the copy constructor only sets the partitioning, it does not copy the actual vector
+    // values
+    mfem::HypreParVector adjoint_load_vector(adjoint_load.trueVec());
+    adjoint_load_vector = adjoint_load.trueVec();
+
+    auto& lin_solver = nonlin_solver_.LinearSolver();
+
+    // By default, use a homogeneous essential boundary condition
+    mfem::HypreParVector adjoint_essential(adjoint_load.trueVec());
+    adjoint_essential = 0.0;
+
+    functional_call_args_[0] = temperature_.trueVec();
+
+    auto [r, drdu] = (*K_functional_)(functional_call_args_, Index<0>{});
+    auto jacobian  = assemble(drdu);
+    auto J_T       = std::unique_ptr<mfem::HypreParMatrix>(jacobian->Transpose());
+
+    // If we have a non-homogeneous essential boundary condition, extract it from the given state
+    if (dual_with_essential_boundary) {
+      dual_with_essential_boundary->initializeTrueVec();
+      adjoint_essential = dual_with_essential_boundary->trueVec();
+    }
+
+    for (const auto& bc : bcs_.essentials()) {
+      bc.eliminateFromMatrix(*J_T);
+      bc.eliminateToRHS(*J_T, adjoint_essential, adjoint_load_vector);
+    }
+
+    lin_solver.SetOperator(*J_T);
+    lin_solver.Mult(adjoint_load_vector, adjoint_temperature_.trueVec());
+
+    adjoint_temperature_.distributeSharedDofs();
+
+    // Reset the equation solver to use the full nonlinear residual operator
+    nonlin_solver_.SetOperator(residual_);
+
+    return adjoint_temperature_;
+  }
+
+  template <int parameter_field>
```

Comment:
What is `parameter_field`? From the code below I see it's an index to a field. My first guess would be "which field index should the sensitivity be wrt", but that's just a guess. A clarification would be good.

In fact, I'm a little confused by the whole function. I can't seem to get the mathematical relation to a sensitivity directly.

Reply:
Yes, I should clarify. This PR is very much a WIP and the documentation/naming schemes need a lot of work. It's probably hard to read at this point, but I wanted some early feedback on the workflow. I'll also be going over it in the meeting today, which will likely be much more understandable.